### ISSUE TITLE
CBMC 5.91 gives unexpected answer on quantified assumptions with unbounded value

### ISSUE THREAD

**Issue by qinheping** [diffblue/cbmc#7891]
CBMC version: 5.91.0
Operating system: macOS Ventura 13.5.2
Exact command line resulting in the issue: 
```
goto-cc main.c
cbmc a.out --smt2
```
What behaviour did you expect: All assertions success.
What happened instead: All assertions fail.
``` c
int main()
{
  unsigned SIZE;
  assume(SIZE >= 5);
  unsigned char *a = malloc(SIZE);

  // clang-format off
  // clang-format would rewrite the "==>" as "== >"
  __CPROVER_assume(__CPROVER_forall { int i; (i>=0 && i<SIZE ) ==> a[i]==1});
  // clang-format on

  assert(a[0]==1);
  assert(a[1]==1);
  assert(a[2]==1);
  assert(a[3]==1);
  assert(a[4]==1);
  return 0;
}
```
In the above code, we assume that all elements in the array `a` equal to `1`, and then check that if the first five elements are equal to 1. 
With the flag `--smt2`, CBMC reports that all checks failed.
With the flag `--cvc4`, CBMC reports that all checks error.
With no flag, CBMC reports that all checks failed.

If we replace `SIZE` by a constant `5` in the assumption, CBMC will give the expected result that all checks success.

**Comment by tautschnig** [comment]
1. There is a problem in the input source code: it should be `__CPROVER_assume`, not just `assume` in line 4. Fixing this makes verification pass as expected with Z3 (which has the necessary quantifier support).
2. When using the propositional back-end, CBMC prints a length warning `warning: ignoring forall ...`, and will over-approximate the result.

As a whole, I would argue that CBMC is behaving as can be currently expected with our limited quantifier instantiation. Still leaving open this issue, but removing the "soundness" label.

**Comment by qinheping** [comment]
> 1. There is a problem in the input source code: it should be `__CPROVER_assume`, not just `assume` in line 4. Fixing this makes verification pass as expected with Z3 (which has the necessary quantifier support).
> 
>     2. When using the propositional back-end, CBMC prints a length warning `warning: ignoring forall ...`, and will over-approximate the result.
> 
> 
> As a whole, I would argue that CBMC is behaving as can be currently expected with our limited quantifier instantiation. Still leaving open this issue, but removing the "soundness" label.

Thank you @tautschnig! 
I tried to use quantifiers in loop contracts and CBMC give the expected result when using Z3
``` c
int main()
{
  unsigned SIZE;
  __CPROVER_assume(SIZE >= 5);
  __CPROVER_assume(SIZE < 64);
  unsigned char *a = malloc(SIZE);

  // clang-format off
  // clang-format would rewrite the "==>" as "== >"
  for(int i = 0; i < SIZE; i++)
  __CPROVER_loop_invariant(i <= SIZE && i >= 0)
  __CPROVER_loop_invariant(__CPROVER_forall{ int j; (j>=0 && j<i) ==> a[j]==1})
  {
      a[i] = 1;
  }
  // clang-format on

  assert(a[0] == 1);
  assert(a[1] == 1);
  assert(a[2] == 1);
  assert(a[3] == 1);
  assert(a[4] == 1);
  return 0;
}
```

However, if we modify the assertion from `assert(a[4] == 1);` to `assert(a[4] == 0);`, CBMC reports that all assertions ERROR instead of that the modified assertion failed.
```
main.c function main
[main.3] line 10 Check loop invariant before entry: SUCCESS
[main.4] line 10 Check loop invariant before entry: ERROR
[main.5] line 10 Check that loop invariant is preserved: ERROR
[main.6] line 10 Check that loop instrumentation was not truncated: ERROR
[main.assigns.1] line 10 Check that i is valid: SUCCESS
[main.assigns.3] line 10 Check that i is assignable: SUCCESS
[main.assigns.5] line 10 Check that i is assignable: SUCCESS
[main.assigns.2] line 14 Check that a[(signed long int)i] is assignable: ERROR
[main.assigns.4] line 14 Check that a[(signed long int)i] is assignable: ERROR
[main.assertion.1] line 18 assertion (signed int)a[(signed long int)0] == 1: ERROR
[main.assertion.2] line 19 assertion (signed int)a[(signed long int)1] == 1: ERROR
[main.assertion.3] line 20 assertion (signed int)a[(signed long int)2] == 1: ERROR
[main.assertion.4] line 21 assertion (signed int)a[(signed long int)3] == 1: ERROR
[main.assertion.5] line 22 assertion (signed int)a[(signed long int)4] == 0: ERROR
```

**Comment by qinheping** [comment]
issue fixed when using z3 4.12
